internal/util/flowgraph/node.go

```diff
@@ -53,58 +53,40 @@ type BaseNode struct {
 	maxParallelism int32
 }
 
-// nodeCtx maintains the running context for a Node in flowgragh
-type nodeCtx struct {
-	node         Node
-	inputChannel chan []Msg
-	downstream   *nodeCtx
-
-	closeCh chan struct{} // notify work to exit
-	closeWg *sync.WaitGroup
-
-	blockMutex sync.RWMutex
-}
-
-// Start invoke Node `Start` method and start a worker goroutine
-func (nodeCtx *nodeCtx) Start() {
-	nodeCtx.node.Start()
-
-	nodeCtx.closeWg.Add(1)
-	go nodeCtx.work()
-}
-
-func (nodeCtx *nodeCtx) Block() {
-	// input node operate function will be blocking
-	if !nodeCtx.node.IsInputNode() {
-		startTs := time.Now()
-		nodeCtx.blockMutex.Lock()
-		if time.Since(startTs) >= blockAllWait {
-			log.Warn("flow graph wait for long time",
-				zap.String("name", nodeCtx.node.Name()),
-				zap.Duration("wait time", time.Since(startTs)))
-		}
-	}
+// manage nodeCtx
+type nodeCtxManager struct {
+	inputNodeCtx *nodeCtx
+	closeWg      *sync.WaitGroup
+	closeOnce    sync.Once
+
+	inputNodeCloseCh chan struct{} // notify input node work to exit
+	ddNodeCloseCh    chan struct{} // notify ddnode and downstream node work to exit
 }
 
-func (nodeCtx *nodeCtx) Unblock() {
-	if !nodeCtx.node.IsInputNode() {
-		nodeCtx.blockMutex.Unlock()
+// NewNodeCtxManager init with the inputNode and fg.closeWg
+func NewNodeCtxManager(nodeCtx *nodeCtx, closeWg *sync.WaitGroup) *nodeCtxManager {
+	return &nodeCtxManager{
+		inputNodeCtx:     nodeCtx,
+		closeWg:          closeWg,
+		inputNodeCloseCh: make(chan struct{}),
+		ddNodeCloseCh:    make(chan struct{}),
 	}
 }
 
-func isCloseMsg(msgs []Msg) bool {
-	if len(msgs) == 1 {
-		return msgs[0].IsClose()
-	}
-	return false
+// Start invoke Node `Start` method and start a worker goroutine
+func (nodeCtxManager *nodeCtxManager) Start() {
+	// in dmInputNode, message from mq to channel, alloc goroutines
```

Comment:
It's betther not mentioning specific node nams in DataNode inside flowgraph node